docs/custom-transfers.md

```diff
@@ -284,6 +349,320 @@ to the stdin of the transfer process:
 On receiving this message the transfer process should clean up and terminate.
 No response is expected.
 
+### Bulk Protocol stages
+
+#### Stage 1: Initiation
+
+Immediately after invoking a bulk transfer process, git-lfs sends initiation
+data to the process over stdin. This tells the process useful information about
+the configuration.
+
+The message will look like this:
+
+```json
+{ "event": "init", "operation": "download", "remote": "origin", "concurrent": true, "concurrenttransfers": 3, "mode": "bulk" }
+```
+
+* `event`: Always `init` to identify this message
+* `operation`: will be `upload` or `download` depending on transfer direction
+* `remote`: The Git remote.  It can be a remote name like `origin` or an URL
+  like `ssh://git.example.com//path/to/repo`.  A standalone transfer agent can
+  use it to determine the location of remote files.
+* `concurrent`: reflects the value of `lfs.bulk.transfer.<name>.concurrent`, in
+  case the process needs to know
+* `concurrenttransfers`: reflects the value of `lfs.concurrenttransfers`, for if
+  the transfer process wants to implement its own concurrency and wants to
+  respect this setting.
+
+The transfer process should use the information it needs from the initiation
+structure, and also perform any one-off setup tasks it needs to do. It should
+then respond on stdout with a confirmation structure, as follows:
+
+```json
+{ "mode": "bulk" }
+```
+
+* `mode`: `bulk` to confirm the bulk mode support. If this value is not
+  specified or set to `basic` - this will switch the transfer to basic mode
+  and all further communications will be performed according to basic protocol.
+
+Or if there was an error:
+
+```json
+{ "error": { "code": 32, "message": "Some init failure message" } }
+```
+
+Responding this way allows to 
+
+#### Stage 2: Bulk Transfer
+After the initiation exchange, git-lfs will send any number of bulk definitions
+to the stdin of the transfer process. Multiple bulk transfers can be processed
+concurrently when the `concurrent` setting is enabled, allowing for efficient
+parallel processing of different bulks across multiple worker processes.
+
+The bulk definition consists of a header, items, and a footer. Each bulk
+definition is sent as a series of JSON messages, each on a single line.
+
+**Important**: If the number of remaining files is less than the configured
+`bulkSize`, git-lfs will automatically flush the pending transfers after a
+100ms timeout to ensure timely processing of incomplete bulks.
+
+
+### 2.1 Bulk Transfer Header
+
+The first message in a bulk transfer is always the bulk header, which defines the
+bulk transfer general bulk data. The header defines the bulk ID and the number of items
+in following bulk transfer. 
+
+It looks like this:
+
+```json
+{ "event": "bulk-header", "oid": "bulk-12345-uuid", "size": 5 }
```

Comment:
I think using `oid` here for a field which isn't a strict Git LFS object ID will be confusing.  May I suggest a field name like `bulk-transfer-id` instead?